Rutefig/reg 473 fix testdecomposedregex in sdk and relayerutils

rutefig opened the pull request:
## Description
Adds regex runtime extraction functionality - previously handled on zk-regex and removed from new compiler fresh implementation - this is used to validate text against regex patterns

Fixes # REG-473

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?
Tested integration directly from the sdk and registry to make sure the overall flow works

- [x] Integration tests from sdk
- [x] Manually testing from registry

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Runtime regex extraction for common email components (addresses, domain, subject, timestamp, message-id, body hash) with standalone and graph-aligned modes.
  * WebAssembly bindings exposing extraction and padding functions to JS.
  * String and byte padding utilities.
  * Public extraction helpers for common patterns and improved error propagation on email-related operations.

* **Chores**
  * Version bumps (Cargo.toml and package.json) and dependency adjustments.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduces a new regex runtime extraction module (Rust + WASM), replaces zk-regex-apis usage, and hardens email handling by validating/piping errors; includes version bumps and dependency updates.
> 
> - **Regex Runtime Extraction (new `src/regex/`)**:
>   - Add `extract`, `patterns`, `padding`, `types`, and WASM bindings providing `extract_substr[_idxes]` and common email pattern helpers.
>   - Support hybrid extraction with optional `NFAGraph`; use `fancy-regex`; define `ExtractionError` and legacy config adapters.
> - **Integration/Refactor**:
>   - Replace `zk_regex_apis` calls with internal `regex` module across `parse_email`, `cryptos`, `circom`, and WASM exports; re-export regex APIs in `lib.rs`.
>   - Add `SerializableRegexPart` and custom serde for `DecomposedRegex.parts`.
> - **Security/Behavior Changes**:
>   - `PaddedEmailAddr::from_email_addr` now validates max length and returns `Result`; propagate in `generate_claim_input`, WASM APIs, and `calculate_account_salt` (now returns `Result`).
> - **Tests**:
>   - Add unit tests for regex extraction, padding, and email length validation/error paths.
> - **Deps/Version**:
>   - Remove `zk-regex-apis`; add `thiserror`, `fancy-regex`; bump crate to `0.4.62-15` and npm to `0.4.66-7`.
> - **Misc**:
>   - `.gitignore`: add `.claude` and `thoughts/`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 4432090d79187ab8c89bb9a7cd9081d6f9d6db69. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

rutefig committed:
chore: migrate regex extraction from zk-regex-apis to local module

Replace the deprecated zk-regex-apis dependency with a new internal
regex module that provides runtime extraction while maintaining
compatibility with the zk-regex-compiler.

Key changes:
- Create modular src/regex/ with extraction, patterns, padding, WASM
- Implement hybrid extraction supporting optional NFAGraph alignment
- Use fancy-regex for full pattern compatibility
- Add custom serde for RegexPart enum serialization
- Remove zk-regex-apis dependency entirely
- Migrate all existing code to use new local regex module

This maintains backward compatibility while modernizing the codebase
to work with the new compiler API.

rutefig committed:
fix: apply clippy suggestions to circom.rs

Remove unused imports and fix code quality issues:
- Remove unused js_sys::Number import
- Remove unused extract_substr_idxes import
- Fix unnecessary clone on Copy type (ProvingFramework)
- Change &Vec<RegexPart> to &[RegexPart] for better idiomatic Rust
- Remove redundant clone() in regex_parts_serde serialization

This reduces compiler warnings and improves code quality.

rutefig committed:
chore: bump version to 0.4.66-5

rutefig committed:
fix: handle legacy decomposed regex struct

rutefig committed:
fix: remove automatic conversion to non capturing group when part is private

rutefig committed:
chore: bump version 0.4.66-7

rutefig commented on src/regex/extract.rs:
```diff
@@ -0,0 +1,301 @@
+            RegexPart::Pattern(p) => {
+                // Wrap in non-capturing group
+                pattern.push_str(p);
+            },
```
In this case I think we should understand if we want to convert automatically or give the control to the user, when we have more feedback on this we can change or not

BornPsych replied:
When zk-regex-compiler generates the NFAGraph, it only counts public patterns as capture groups. This is not configurable in circuits, should we give out the control to users for that, since this is the only way to do that? or am I missing something

rutefig replied:
I think there's some confusion here. Let me clarify:
- The circuit correctly only extracts PublicPatterns - that's working as intended.
- The actual problem is:
If users write Pattern parts with regular parentheses (...), those create numbered capture groups in the regex itself. For example:

```
// This is WRONG - creates capture group 1:
Pattern("(a|b)prefix:")           // ← capture group 1
PublicPattern("[0-9]+", 10)       // ← capture group 2 (but should be 1)

// This is CORRECT - no extra capture group:
Pattern("(?:a|b)prefix:")         // ← non-capturing group (group 0, ignored)
PublicPattern("[0-9]+", 10)       // ← capture group 1
```

**The real issue:** Users need to write `(?:...)` (non-capturing) instead of `(...)` (capturing) in `Pattern` parts. Otherwise they create unwanted numbered groups that mess up the `PublicPattern` numbering.

Current design: Both old and new compilers give users full control - they must write `(?:...)` explicitly.
The bug in the code: The misleading comment at line 152 that says `// Wrap in non-capturing group` but the code doesn't actually wrap. - will fix this

So the control is already with users - they choose when to use `(?:...)`. The question is: should we automatically wrap Pattern parts in `(?:...)` to prevent this mistake?

BornPsych replied:
I think we should check in our implementation and automatically wrap pattern parts to prevent the mistake

rutefig replied:
Yes I agree, so now the questions comes to where should we do this wrapping?

We have 3 options: `sdk`, `relayer-utils` or `zk-regex`

On my opinion since we have the definition of private and public parts on `zk-regex` I think this would make sense to be wrapped there, but since this is not the case and I'm not sure to have the whole context would be good to have more thoughts on this from you and @shreyas-londhe 
This would be the place: https://github.com/zkemail/zk-regex/pull/92/files#diff-002f72d74839fa95c8728e26bd644cc65d6bd86b9e3a4f992e4062028921e3aeR33

shreyas-londhe replied:
This is a valid point, thank for pointing it out @rutefig. 

Yes, it is misleading and gives users full control unkowingly. The best thing to do over here is to analyse the `Pattern` regex and either throw an error which indicates that an capture group is used in a non-capturing part or we programatically change that capture into a non-capture.

rutefig replied:
This is being solved on this [PR](https://github.com/zkemail/zk-regex/pull/109)

rutefig commented:
@coderabbitai review

rutefig commented on src/regex/wasm.rs:
```diff
@@ -0,0 +1,220 @@
+        println!("result: {:?}", result);
+        assert_eq!(result.unwrap(), vec!["alice@example.com"]);
+    }
+}
```
Ah this is true, but I think we can leave this for later. I left the tests anyways and I believe it is worth adding `wasm-bindgen-test` later on

rutefig replied:
removed this for now

BornPsych committed:
chore: doc test fix and cargo fmt

rutefig committed:
fix: wrong comments on wrapping non capture groups since we don't do that

rutefig committed:
fix: convert bare groups to non capturing groups to ensure predictable numbering

rutefig commented on src/regex/extract.rs:
```diff
@@ -0,0 +1,301 @@
+    for part in &config.parts {
+        match part {
+            RegexPart::Pattern(p) => {
+                // Wrap in non-capturing group
+                pattern.push_str(p);
+            },
+            RegexPart::PublicPattern((p, _max_bytes)) => {
+                public_count += 1;
+                // Wrap in capturing group
+                pattern.push_str(&format!("({})", p).as_str());
+            }
+        };
```
Looking at the previous implementation on zk-regex apis we had [this line](https://github.com/zkemail/zk-regex/blob/4fb1bda1af71123f57f4877c29ebcb9004266fd8/packages/apis/src/extract_substrs.rs#L58) that converts all bare capturing groups inside of the `PublicPattern` to non-capturing groups. So we should maintain this behaviour instead of erroring out. Just [pushed a fix](https://github.com/zkemail/relayer-utils/pull/34/commits/7659dc918be957a8bc4208a8499d8273c2bb9a9c) to address this.

rutefig committed:
fix: add validation to prevent email address truncation in PaddedEmailAddr

Add explicit length validation in PaddedEmailAddr::from_email_addr() to
reject email addresses exceeding 256 bytes. This prevents a critical
security issue where different email addresses with identical first 256
bytes could produce the same cryptographic commitments and account salts
due to silent truncation.

Changes:
- Change PaddedEmailAddr::from_email_addr() to return Result<Self>
- Add validation to reject emails exceeding MAX_EMAIL_ADDR_BYTES (256)
- Update calculate_account_salt() to return Result<String>
- Update all WASM bindings with proper error handling
- Add comprehensive test coverage for validation scenarios
- Document truncation behavior in pad_string() with security notes

Breaking change: This changes the public API signature of
from_email_addr() and calculate_account_salt() to return Results.

Addresses CodeRabbit security review comment on PR #34

rutefig commented on src/regex/padding.rs:
```diff
@@ -0,0 +1,50 @@
+///
+/// # Returns
+/// A vector of bytes padded to target_len with zeros
+pub fn pad_string(s: &str, target_len: usize) -> Vec<u8> {
+    let mut padded = s.as_bytes().to_vec();
+    padded.resize(target_len, 0);
+    padded
+}
```
Pushed a fix for this

rutefig committed:
fix: properly gate wasm modules and remove unnecessary import

rutefig commented on src/regex/wasm.rs:
```diff
@@ -0,0 +1,220 @@
+use js_sys::Array;
+use wasm_bindgen::prelude::*;
+use serde_wasm_bindgen::{from_value, to_value};
+
+use crate::regex::types::{DecomposedRegexConfig, ExtractionError};
+use crate::regex::extract::{extract_substr_idxes, extract_substr};
+use crate::regex::patterns::*;
+use crate::regex::padding::pad_string;
+
```
Fixed

rutefig committed:
chore: remove wasm bindgen tests for now

rutefig committed:
chore: added test for extract_substr with diff ways of defining subject all pattern

rutefig commented:
@coderabbitai review

BornPsych reviewed (approved):
LGTM

rutefig committed:
chore: bump version
